### PR TITLE
 feat: automatic session implementation

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description

This PR implements automatic session management.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details

- `SessionConfiguration` - A class designed to manage session settings.
   - `sessionTimeoutInMillis` - Specifies the session timeout duration in milliseconds, with a default of 5 minutes.
   - When `automaticSessionTracking` is enabled (by default), `SessionManager` registers following app lifecycle observers to update the last activity time and manage session timeouts.
   
      - `UIApplication.didEnterBackgroundNotification`
      - `UIApplication.willTerminateNotification`
      - `UIApplication.willEnterForegroundNotification`

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

## How to test?

- Run the app (automatic session enabled by default) and make any event to check the context section for added details.

### Sample Test Scenarios:

- Run the app, make any event and check the session details.
- Run the app, make multiple events, and check the session details.
- Run the app, make any event, stop the session, make another event and check the session details added or not.
- Run the app, make any event, perform reset action, make another event and check the session details.
- Run the app, make any event, relaunch the app, make another event and check the session details.
- Run the app, start a manual session, make an event and check the session details.
- Run the app, start a manual session,  relaunch the app, make an event and check the session details.

First event after a session starts,
```Json
{
    "anonymousId": "FC05A8C6-8FF3-4419-850F-EDF80F857DBB",
    "channel": "mobile",
    "context": {
        "SK1": {
            "Key1": "Value1"
        },
        "SK2": [
            "value1",
            "value2"
        ],
        "SK3": "Value3",
        "SK4": 1234,
        "SK5": 5678.9,
        "SK6": true,
        "app": {
            "build": "1",
            "name": "AnalyticsApp",
            "namespace": "com.ska.AnalyticsApp",
            "version": "1.0"
        },
        "device": {
            "id": "A952F66B-FD41-4B00-AC66-8EA3857F4330",
            "manufacturer": "Apple",
            "model": "arm64",
            "name": "iPhone 16 Pro",
            "type": "iOS"
        },
        "externalId": [
            {
                "id": "12791",
                "type": "idCardNumber"
            }
        ],
        "library": {
            "name": "rudder-ios-library",
            "version": "1.31.0"
        },
        "locale": "en-IN",
        "network": {
            "bluetooth": true,
            "cellular": false,
            "wifi": true
        },
        "os": {
            "name": "iOS",
            "version": "18.2"
        },
        "screen": {
            "density": 3,
            "height": 874,
            "width": 402
        },
        "sessionId": 1741008868,
        "sessionStart": true,
        "timezone": "Asia/Kolkata"
    },
    "event": "Track at 2025-03-03 13:34:36 +0000",
    "integrations": {
        "All": true,
        "Amplitude": true,
        "CleverTap": false
    },
    "messageId": "E0B4E9C1-1935-4027-925C-AF15C812CE41",
    "originalTimeStamp": "2025-03-03T13:34:36.118Z",
    "properties": {
        "key": "value"
    },
    "receivedAt": "2025-03-03T13:34:37.120Z",
    "request_ip": "106.51.148.108",
    "rudderId": "9f0451ea-0175-4dae-a1ee-9cc0b90d9bb1",
    "sentAt": "2025-03-03T13:34:36.138Z",
    "type": "track"
}
```

Next event after a session starts,
```Json
{
    "anonymousId": "FC05A8C6-8FF3-4419-850F-EDF80F857DBB",
    "channel": "mobile",
    "context": {
        "SK": {
            "Key1": "Value1"
        },
        "app": {
            "build": "1",
            "name": "AnalyticsApp",
            "namespace": "com.ska.AnalyticsApp",
            "version": "1.0"
        },
        "device": {
            "id": "A952F66B-FD41-4B00-AC66-8EA3857F4330",
            "manufacturer": "Apple",
            "model": "arm64",
            "name": "iPhone 16 Pro",
            "type": "iOS"
        },
        "externalId": [
            {
                "id": "12791",
                "type": "idCardNumber"
            }
        ],
        "library": {
            "name": "rudder-ios-library",
            "version": "1.31.0"
        },
        "locale": "en-IN",
        "network": {
            "bluetooth": true,
            "cellular": false,
            "wifi": true
        },
        "os": {
            "name": "iOS",
            "version": "18.2"
        },
        "screen": {
            "density": 3,
            "height": 874,
            "width": 402
        },
        "sessionId": 1741008868,
        "timezone": "Asia/Kolkata"
    },
    "event": "Analytics Screen",
    "integrations": {
        "All": true,
        "Facebook": false
    },
    "messageId": "70F486C0-1551-4707-873C-CD45C70B76C9",
    "originalTimeStamp": "2025-03-03T13:34:51.656Z",
    "properties": {
        "key": "value",
        "name": "Analytics Screen"
    },
    "receivedAt": "2025-03-03T13:34:59.140Z",
    "request_ip": "106.51.148.108",
    "rudderId": "9f0451ea-0175-4dae-a1ee-9cc0b90d9bb1",
    "sentAt": "2025-03-03T13:34:58.975Z",
    "type": "screen"
}
```

## Maintainers Checklist

- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

## Screenshots (if applicable)

Simulator Screenshot 2025-03-01

### CODE REVIEW

**Review by ChryssaAliferi — APPROVED** [review]
Looks good to me 👀 
Let's make sure we absolutely test the functionality.

File: Analytics/Analytics/SessionManagement/SessionManager.swift [hunk 1]
```diff
@@ -21,32 +25,93 @@ enum SessionType {
...
-        self.startSession(id: SessionManager.generatedSessionId, shouldUpdateType: false)
+        self.startSession(id: SessionManager.generatedSessionId, type: self.sessionType)
+    }
+    
+    func ensureAutomaticSession() {
+        guard self.sessionCofiguration.automaticSessionTracking else { return }
+        if self.sessionId == nil || self.sessionType == .manual || self.isSessionTimedOut {
+            self.startSession(id: Self.generatedSessionId, type: .automatic)
+        }
+    }
```
Comment (1abhishekpandey) [inline Analytics/Analytics/SessionManagement/SessionManager.swift]: Although the name of the method is `ensure...` we are doing more than that. This method seems to handle the logic of the automatic session tracking.

Please change the name of this method to something more meaningful, e.g., `startAutomaticSessionIfNeeded()`.

Objective: Method name should match it's intended behaviour. Otherwise it would be hard to maintain and understand it's purpose.

File: Analytics/Analytics/SessionManagement/SessionManager.swift [hunk 2]
```diff
@@ -21,32 +25,93 @@ enum SessionType {
...
+    deinit {
+        self.detachObservers()
+    }
+}
+
+// MARK: - Observers
+
+extension SessionManager {
+    
+    func attachObservers() {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/SessionManagement/SessionManager.swift:83]: I think this observer is needed to track the application foreground and background activities. In future, I think it will be replaced with the observer pattern i.e., this class should only subscribe to those activities and shouldn't implement them here. (Refer to the Android code for more details.)

Therefore, I would suggest either adding a proper TODO, explaining that we need to move this, or, do it now by extracting out the notification related code.

File: Analytics/Analytics/SessionManagement/SessionManager.swift [hunk 3]
```diff
@@ -21,32 +25,93 @@ enum SessionType {
...
     
-    func startSession(id: UInt64, type: SessionType = SessionConstants.defaultSessionType, shouldUpdateType: Bool = true) {
+    func startSession(id: UInt64, type: SessionType) {
         self.updateSessionStart(isSessionStrat: true)
-        if shouldUpdateType {
-            self.updateSessionType(type: type)
-        }
+        self.updateSessionType(type: type)
         self.updateSessionId(id: id)
+        self.sessionType == .automatic ? self.attachObservers() : self.detachObservers()
```
Comment (1abhishekpandey) [inline Analytics/Analytics/SessionManagement/SessionManager.swift:50]: Great use of `observers`, specifically the detaching portion.
But please ensure we have the parity between Android and iOS.

**Review by vgupta98 — APPROVED** [review]

File: Analytics/Analytics/SessionManagement/SessionManager.swift (hunk 1 above)
Comment (1abhishekpandey) [reply Analytics/Analytics/SessionManagement/SessionManager.swift]: Also, handle the edge case of automatic session tracking, i.e., if previously session tracking is enabled and in the next app launch if it's disabled then session tracking should be stopped.

File: Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift [hunk 4]
```diff
@@ -10,83 +10,225 @@ import XCTest
...
+            when("The app moves to the background") {
+                NotificationCenter.default.post(name: backgroundNotification, object: nil)
+                
+                then("The session's last activity time should update") {
+                    XCTAssertFalse(manager.isSessionTimedOut, "Session should not immediately be timed out after backgrounding")
+                }
+            }
+            
+            when("The app comes back to the foreground before timeout") {
+                sleep(1) // Simulating time passing (depends on session timeout duration)
```
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift]: If this `sleep()` increases the test execution time even by a second, then I would propose not to have this test at all. Having a delay in the unit test is not a good approach in my understanding. Try another way if possible, else simply remove the test.

Comment (1abhishekpandey) [reply Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift]: Remove all the use of `sleep` from this test class.

Comment (1abhishekpandey) [reply Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift]: Reason: If test execution is delayed, then it'll become hard to maintain. Ideally, we should aim for the unit test to execute as quickly as possible.

File: Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift [hunk 5]
```diff
@@ -10,83 +10,225 @@ import XCTest
...
+                
+                then("A new session should be started if timed out") {
+                    XCTAssertTrue(manager.isSessionTimedOut, "Session should be timed out")
+                    XCTAssertNotEqual(manager.sessionId, initialSessionId, "A new session should be created after timeout")
+                }
+            }
+        }
+    }
+    
+    func testAttachObservers_RegistersNotifications() {
```
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift:182]: Let's add a TODO to indicate this need to be moved out of the session tracking.

Session class shouldn't be responsible for observer registering. 

Also, in Android, we have implemented the observer pattern. So, as per my understanding we are going to follow that. And once we follow that, we wouldn't need these tests.

File: Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift [hunk 6]
```diff
@@ -10,83 +10,225 @@ import XCTest
...
+                then("Observers should be registered") {
+                    XCTAssertNotNil(manager.backgroundObserver, "Background observer should be registered")
+                    XCTAssertNotNil(manager.foregroundObserver, "Foreground observer should be registered")
+                    XCTAssertNotNil(manager.terminateObserver, "Terminate observer should be registered")
+                }
+            }
+        }
+    }
+    
+    func testDetachObservers_RemovesNotifications() {
```
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift:202]: Add a TODO to remove for this test as well.

File: Analytics/Analytics/SessionManagement/SessionManager.swift (hunk 1 above)
Comment (1abhishekpandey) [reply Analytics/Analytics/SessionManagement/SessionManager.swift]: If possible, let's add a unit test case for the above edge case.

**Review by 1abhishekpandey — CHANGES_REQUESTED** [review]

File: Analytics/Analytics/SessionManagement/SessionManager.swift (hunk 1 above)
Comment (SKannaniOS) [reply Analytics/Analytics/SessionManagement/SessionManager.swift]: I updated the method definition to cover the edge case & added test case for it.

File: Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift (hunk 4 above)
Comment (SKannaniOS) [reply Analytics/AnalyticsTests/SessionManagementTests/SessionManagerTests.swift]: I have updated the test case by eliminating the need of `sleep()` calls.

File: Analytics/Analytics/SessionManagement/SessionManager.swift (hunk 3 above)
Comment (vgupta98) [reply Analytics/Analytics/SessionManagement/SessionManager.swift:50]: Let's refactor this on Android side as well. Detaching the observers when session is converted to manual is a good idea. I can add it in the session tracking plugin refactor PR in future.

File: Analytics/Analytics/SessionManagement/SessionManager.swift (hunk 1 above)
Comment (1abhishekpandey) [reply Analytics/Analytics/SessionManagement/SessionManager.swift]: @SKannaniOS Can you please address [this](https://github.com/rudderlabs/swift-sdk-poc/pull/43#discussion_r1980685050) comment? Or feel free to reach out if you have any questions!

**Review by 1abhishekpandey — APPROVED** [review]
